File: kalitech-engine/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainCreateHelper.js
// FILE: resources/kalitech/builtin/terrain/TerrainCreateHelper.js
"use strict";

const {isObj, num, i32} = require("./TerrainTypes.js");
const {TerrainHeights} = require("./TerrainHeights.js");
const {TerrainInstance} = require("./TerrainInstance.js");

class TerrainCreateHelper {
    constructor(engine, terrNative, heightsApi, physicsApi, proxies) {
        this.engine = engine;
        this.terr = terrNative;
        this.heights = heightsApi;      // TerrainHeights instance
        this.phys = physicsApi;         // TerrainPhysics instance
        this.p = proxies;               // proxy methods { terrain, terrainHeights, quad, plane, ... }
    }

    create(cfg) {
        const c = isObj(cfg) ? cfg : {};
        const kind = String(c.kind || "terrain");
        const attachFlag = (c.attach === undefined) ? true : !!c.attach;

        const materialH = c.material;
        const uvCfg = c.uv;
        const lodCfg = c.lod;
        const physCfg = c.physics;

        const scaleCfg = isObj(c.scale) ? c.scale : null;
        const xz = scaleCfg ? num(scaleCfg.xz, num(c.xzScale, 1.0)) : num(c.xzScale, 1.0);
        const y = scaleCfg ? num(scaleCfg.y, num(c.yScale, num(c.heightScale, 1.0))) : num(c.yScale, num(c.heightScale, 1.0));

        // ------------------------------------------------------------
        // post: apply material/uv/lod/scale and RETURN TerrainInstance
        // ------------------------------------------------------------
        const post = (surfaceOrBundle) => {
            const surface = (surfaceOrBundle && surfaceOrBundle.surface) ? surfaceOrBundle.surface : surfaceOrBundle;
            if (!surface) throw new Error("[TERR] create(): native returned null surface");

            if (materialH != null) this.p.material(surface, materialH);
            if (uvCfg != null) this.p.uv(surface, uvCfg);
            if (lodCfg != null) this.p.lod(surface, lodCfg);

            // behavior: only non-plane/quad get post-scale here (as in original)
            if (kind !== "plane" && kind !== "quad") {
                if (Number.isFinite(xz) && xz !== 1.0) this.p.scale(surface, xz, {yScale: y});
                else if (Number.isFinite(y) && y !== 1.0) this.p.scale(surface, 1.0, {yScale: y});
            }

            // IMPORTANT: return object-mode instance (mirrors Java API via proxies)
            return new TerrainInstance(this.p, surface);
        };

        // ------------------------------------------------------------
        // plane
        // ------------------------------------------------------------
        if (kind === "plane") {
            const planeCfg = Object.assign({}, isObj(c.plane) ? c.plane : {}, {
                name: c.name,
                attach: attachFlag,
                physics: physCfg,
            });
            return post(this.p.plane(planeCfg));
        }

        // ------------------------------------------------------------
        // quad
        // ------------------------------------------------------------
        if (kind === "quad") {
            const quadCfg = Object.assign({}, isObj(c.quad) ? c.quad : {}, {
                name: c.name,
                attach: attachFlag,
                physics: physCfg,
            });
            return post(this.p.quad(quadCfg));
        }

        // ------------------------------------------------------------
        // heightmap
        // ------------------------------------------------------------
        if (kind === "heightmap") {
            const tcfg = Object.assign({}, isObj(c.terrain) ? c.terrain : {}, {
                name: c.name,
                attach: attachFlag,
                physics: physCfg,
            });
            if (c.heightmap && !tcfg.heightmap) tcfg.heightmap = c.heightmap;
            if (tcfg.heightScale == null && Number.isFinite(y)) tcfg.heightScale = y;
            if (tcfg.xzScale == null && Number.isFinite(xz)) tcfg.xzScale = xz;

            // native creation already handles physics via p.terrain wrapper
            return post(this.p.terrain(tcfg));
        }

        // ------------------------------------------------------------
        // noise: generate heights -> create terrainHeights -> apply post -> build physics AFTER post-scale
        // ------------------------------------------------------------
        if (kind === "noise") {
            const noise = isObj(c.noise) ? c.noise : {};
            const type = String(noise.type || "perlin");

            const size = i32((isObj(c.terrain) ? c.terrain.size : c.size), i32(noise.size, 513)) || 513;
            const patchSize = i32((isObj(c.terrain) ? c.terrain.patchSize : c.patchSize), 65) || 65;

            TerrainHeights.validateTerrainDims(size, patchSize);

            const raw = (type === "ridged")
                ? this.heights.ridged(Object.assign({}, noise, {size}))
                : this.heights.perlin(Object.assign({}, noise, {size}));

            const normalize = (noise.normalize === undefined) ? true : !!noise.normalize;

            const out = new Float32Array(raw.length);
            if (normalize) {
                for (let i = 0; i < raw.length; i++) out[i] = (raw[i] * 2.0 - 1.0) * y;
            } else {
                for (let i = 0; i < raw.length; i++) out[i] = raw[i] * y;
            }

            const tcfg0 = isObj(c.terrain) ? c.terrain : {};
            const tcfgNoPhys = Object.assign({}, tcfg0, {
                name: c.name,
                size,
                patchSize,
                heights: out,
                attach: attachFlag,
            });

            // IMPORTANT: create WITHOUT physics first (dynamicMesh later)
            let surface = this.terr.terrainHeights(tcfgNoPhys);

            // apply post (material/uv/lod/scale)
            const inst = post(surface);

            // build physics AFTER post-scale, but do not change returned type (instance)
            // (physics attachment is side-effect; the actual body handle lives in engine registries)
            if (physCfg != null) this.phys.withBody(this.terr, inst.surface, physCfg, "dynamicMesh");

            return inst;
        }

        // ------------------------------------------------------------
        // heights: accept array-like -> f32, infer size if possible, validate dims, create terrainHeights
        // ------------------------------------------------------------
        if (kind === "heights") {
            const heightsIn = c.heights;
            if (!heightsIn) throw new Error("[TERR] create(kind='heights'): cfg.heights is required");

            const heights = TerrainHeights.toFloat32Array(heightsIn);

            // allow either cfg.terrain.{size,patchSize} or top-level size/patchSize
            const tcfg0 = isObj(c.terrain) ? c.terrain : {};
            const size = i32(tcfg0.size, i32(c.size, 0)) || TerrainHeights.inferSizeFromHeights(heights);
            const patchSize = i32(tcfg0.patchSize, i32(c.patchSize, 0));

            if (size > 0) {
                TerrainHeights.validateTerrainDims(size, patchSize || 0);
                TerrainHeights.assertHeightsMatchSize(heights, size, "create(kind='heights')");
            }

            const tcfgNoPhys = Object.assign({}, tcfg0, {
                name: c.name,
                heights,
                // keep explicit size/patchSize if known
                size: size || undefined,
                patchSize: patchSize || undefined,
                attach: attachFlag,
            });

            let surface = this.terr.terrainHeights(tcfgNoPhys);

            const inst = post(surface);

            // IMPORTANT: build physics AFTER post-scale (dynamicMesh)
            if (physCfg != null) this.phys.withBody(this.terr, inst.surface, physCfg, "dynamicMesh");

            return inst;
        }

        // ------------------------------------------------------------
        // default: treat as heightmap/terrain cfg
        // ------------------------------------------------------------
        return post(this.p.terrain(Object.assign({}, isObj(c.terrain) ? c.terrain : c)));
    }
}

module.exports = {TerrainCreateHelper};

File: modulesSrc/TerrainModule/src/main/resources/kalitech/engine/modules/Terrain/helpers/TerrainCreateHelper.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaTableMerge = luaRuntime.LuaTableMerge
local lua_require_result_0 = require("./TerrainTypes.lua")
isObj = lua_require_result_0.isObj
num = lua_require_result_0.num
i32 = lua_require_result_0.i32
local lua_require_result_1 = require("./TerrainHeights.lua")
TerrainHeights = lua_require_result_1.TerrainHeights
local lua_require_result_2 = require("./TerrainInstance.lua")
TerrainInstance = lua_require_result_2.TerrainInstance
TerrainCreateHelper = LuaClass()
TerrainCreateHelper.name = "TerrainCreateHelper"
function TerrainCreateHelper.prototype.lua_constructor(self, engine, terrNative, heightsApi, physicsApi, proxies)
    self.engine = engine
    self.terr = terrNative
    self.heights = heightsApi
    self.phys = physicsApi
    self.p = proxies
end
function TerrainCreateHelper.prototype.create(self, cfg)
    local lua_isObj_result_3
    if isObj(_G, cfg) then
        lua_isObj_result_3 = cfg
    else
        lua_isObj_result_3 = {}
    end
    local c = lua_isObj_result_3
    local kind = tostring(c.kind or "terrain")
    local lua_temp_4
    if c.attach == nil then
        lua_temp_4 = true
    else
        lua_temp_4 = not not c.attach
    end
    local attachFlag = lua_temp_4
    local materialH = c.material
    local uvCfg = c.uv
    local lodCfg = c.lod
    local physCfg = c.physics
    local lua_isObj_result_5
    if isObj(_G, c.scale) then
        lua_isObj_result_5 = c.scale
    else
        lua_isObj_result_5 = nil
    end
    local scaleCfg = lua_isObj_result_5
    local lua_scaleCfg_6
    if scaleCfg then
        lua_scaleCfg_6 = num(
            _G,
            scaleCfg.xz,
            num(_G, c.xzScale, 1)
        )
    else
        lua_scaleCfg_6 = num(_G, c.xzScale, 1)
    end
    local xz = lua_scaleCfg_6
    local lua_scaleCfg_7
    if scaleCfg then
        lua_scaleCfg_7 = num(
            _G,
            scaleCfg.y,
            num(
                _G,
                c.yScale,
                num(_G, c.heightScale, 1)
            )
        )
    else
        lua_scaleCfg_7 = num(
            _G,
            c.yScale,
            num(_G, c.heightScale, 1)
        )
    end
    local y = lua_scaleCfg_7
    local function post(lua_, surfaceOrBundle)
        local lua_temp_8
        if surfaceOrBundle and surfaceOrBundle.surface then
            lua_temp_8 = surfaceOrBundle.surface
        else
            lua_temp_8 = surfaceOrBundle
        end
        local surface = lua_temp_8
        if not surface then
            error(
                LuaConstruct(Error, "[TERR] create(): native returned null surface"),
                0
            )
        end
        if materialH ~= nil then
            self.p:material(surface, materialH)
        end
        if uvCfg ~= nil then
            self.p:uv(surface, uvCfg)
        end
        if lodCfg ~= nil then
            self.p:lod(surface, lodCfg)
        end
        if kind ~= "plane" and kind ~= "quad" then
            if LuaNumberIsFinite(xz) and xz ~= 1 then
                self.p:scale(surface, xz, {yScale = y})
            elseif LuaNumberIsFinite(y) and y ~= 1 then
                self.p:scale(surface, 1, {yScale = y})
            end
        end
        return LuaConstruct(TerrainInstance, self.p, surface)
    end
    if kind == "plane" then
        local lua_temp_10 = {}
        local lua_isObj_result_9
        if isObj(_G, c.plane) then
            lua_isObj_result_9 = c.plane
        else
            lua_isObj_result_9 = {}
        end
        local planeCfg = LuaTableMerge(lua_temp_10, lua_isObj_result_9, {name = c.name, attach = attachFlag, physics = physCfg})
        return post(
            _G,
            self.p:plane(planeCfg)
        )
    end
    if kind == "quad" then
        local lua_temp_12 = {}
        local lua_isObj_result_11
        if isObj(_G, c.quad) then
            lua_isObj_result_11 = c.quad
        else
            lua_isObj_result_11 = {}
        end
        local quadCfg = LuaTableMerge(lua_temp_12, lua_isObj_result_11, {name = c.name, attach = attachFlag, physics = physCfg})
        return post(
            _G,
            self.p:quad(quadCfg)
        )
    end
    if kind == "heightmap" then
        local lua_temp_14 = {}
        local lua_isObj_result_13
        if isObj(_G, c.terrain) then
            lua_isObj_result_13 = c.terrain
        else
            lua_isObj_result_13 = {}
        end
        local tcfg = LuaTableMerge(lua_temp_14, lua_isObj_result_13, {name = c.name, attach = attachFlag, physics = physCfg})
        if c.heightmap and not tcfg.heightmap then
            tcfg.heightmap = c.heightmap
        end
        if tcfg.heightScale == nil and LuaNumberIsFinite(y) then
            tcfg.heightScale = y
        end
        if tcfg.xzScale == nil and LuaNumberIsFinite(xz) then
            tcfg.xzScale = xz
        end
        return post(
            _G,
            self.p:terrain(tcfg)
        )
    end
    if kind == "noise" then
        local lua_isObj_result_15
        if isObj(_G, c.noise) then
            lua_isObj_result_15 = c.noise
        else
            lua_isObj_result_15 = {}
        end
        local noise = lua_isObj_result_15
        local lua_type = tostring(noise.type or "perlin")
        local lua_i32_18 = i32
        local lua_G_17 = _G
        local lua_isObj_result_16
        if isObj(_G, c.terrain) then
            lua_isObj_result_16 = c.terrain.size
        else
            lua_isObj_result_16 = c.size
        end
        local size = lua_i32_18(
            lua_G_17,
            lua_isObj_result_16,
            i32(_G, noise.size, 513)
        ) or 513
        local lua_i32_21 = i32
        local lua_G_20 = _G
        local lua_isObj_result_19
        if isObj(_G, c.terrain) then
            lua_isObj_result_19 = c.terrain.patchSize
        else
            lua_isObj_result_19 = c.patchSize
        end
        local patchSize = lua_i32_21(lua_G_20, lua_isObj_result_19, 65) or 65
        TerrainHeights:validateTerrainDims(size, patchSize)
        local lua_temp_22
        if lua_type == "ridged" then
            lua_temp_22 = self.heights:ridged(LuaTableMerge({}, noise, {size = size}))
        else
            lua_temp_22 = self.heights:perlin(LuaTableMerge({}, noise, {size = size}))
        end
        local raw = lua_temp_22
        local lua_temp_23
        if noise.normalize == nil then
            lua_temp_23 = true
        else
            lua_temp_23 = not not noise.normalize
        end
        local normalize = lua_temp_23
        local out = {}
        if normalize then
            do
                local i = 0
                while i < KLength(raw) do
                    KSetIndex(out, i, (KIndex(raw, i) * 2 - 1) * y)
                    i = i + 1
                end
            end
        else
            do
                local i = 0
                while i < KLength(raw) do
                    KSetIndex(out, i, KIndex(raw, i) * y)
                    i = i + 1
                end
            end
        end
        local lua_isObj_result_24
        if isObj(_G, c.terrain) then
            lua_isObj_result_24 = c.terrain
        else
            lua_isObj_result_24 = {}
        end
        local tcfg0 = lua_isObj_result_24
        local tcfgNoPhys = LuaTableMerge({}, tcfg0, {
            name = c.name,
            size = size,
            patchSize = patchSize,
            heights = out,
            attach = attachFlag
        })
        local surface = self.terr:terrainHeights(tcfgNoPhys)
        local inst = post(_G, surface)
        if physCfg ~= nil then
            self.phys:withBody(self.terr, inst.surface, physCfg, "dynamicMesh")
        end
        return inst
    end
    if kind == "heights" then
        local heightsIn = c.heights
        if not heightsIn then
            error(
                LuaConstruct(Error, "[TERR] create(kind='heights'): cfg.heights is required"),
                0
            )
        end
        local heights = TerrainHeights:toFloatArray(heightsIn)
        local lua_isObj_result_25
        if isObj(_G, c.terrain) then
            lua_isObj_result_25 = c.terrain
        else
            lua_isObj_result_25 = {}
        end
        local tcfg0 = lua_isObj_result_25
        local size = i32(
            _G,
            tcfg0.size,
            i32(_G, c.size, 0)
        ) or TerrainHeights:inferSizeFromHeights(heights)
        local patchSize = i32(
            _G,
            tcfg0.patchSize,
            i32(_G, c.patchSize, 0)
        )
        if size > 0 then
            TerrainHeights:validateTerrainDims(size, patchSize or 0)
            TerrainHeights:assertHeightsMatchSize(heights, size, "create(kind='heights')")
        end
        local tcfgNoPhys = LuaTableMerge({}, tcfg0, {
            name = c.name,
            heights = heights,
            size = size or nil,
            patchSize = patchSize or nil,
            attach = attachFlag
        })
        local surface = self.terr:terrainHeights(tcfgNoPhys)
        local inst = post(_G, surface)
        if physCfg ~= nil then
            self.phys:withBody(self.terr, inst.surface, physCfg, "dynamicMesh")
        end
        return inst
    end
    local lua_G_30 = _G
    local lua_self_28 = self.p
    local lua_self_28_terrain_29 = lua_self_28.terrain
    local lua_temp_27 = {}
    local lua_isObj_result_26
    if isObj(_G, c.terrain) then
        lua_isObj_result_26 = c.terrain
    else
        lua_isObj_result_26 = c
    end
    return post(
        lua_G_30,
        lua_self_28_terrain_29(
            lua_self_28,
            LuaTableMerge(lua_temp_27, lua_isObj_result_26)
        )
    )
end
M = {TerrainCreateHelper = TerrainCreateHelper}

return M
